File: samples/cpp/neural_networks/sources/daal_lenet.cpp
/* file: daal_lenet.cpp */
/*******************************************************************************
* Copyright 2017-2019 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*
* License:
* http://software.intel.com/en-us/articles/intel-sample-source-code-license-agr
* eement/
*******************************************************************************/

/*
!  Content:
!    C++ example of neural network training and scoring with LeNet topology
!******************************************************************************/

#include "service.h"
#include "daal_lenet.h"
#include "image_dataset.h"

/* Trains AlexNet with given dataset reader */
prediction::ModelPtr train(DatasetReader *reader);

/* Tests AlexNet with given dataset reader */
float test(prediction::ModelPtr predictionModel, DatasetReader *reader);

/* Constructs the optimization solver with given learning rate */
SharedPtr<optimization_solver::adagrad::Batch<float> > getOptimizationSolver(float learningRate);

size_t trainDatasetObjectsNum = 55000;
size_t testDatasetObjectsNum  = 10000;
float top1ErrorRateThreshold  = 0.97;
const size_t maxIteraions     = 10;
const float learningRate      = 0.01;
const size_t batchSize        = 64;

const std::string defaultDatasetsPath = "./data";
const std::string datasetFileNames[] =
{
    "train-images-idx3-ubyte",
    "train-labels-idx1-ubyte",
    "t10k-images-idx3-ubyte",
    "t10k-labels-idx1-ubyte"
};

int main(int argc, char *argv[])
{
    std::string userDatasetsPath = getUserDatasetPath(argc, argv);
    std::string datasetsPath = selectDatasetPathOrExit(
        defaultDatasetsPath, userDatasetsPath, datasetFileNames, 4);

    printf("Data loading started... \n");

    /* Form path to the training and testing datasets */
    std::string trainDatasetImages = datasetsPath + "/" + datasetFileNames[0];
    std::string trainDatasetLabels = datasetsPath + "/" + datasetFileNames[1];
    std::string testDatasetImages  = datasetsPath + "/" + datasetFileNames[2];
    std::string testDatasetLabels  = datasetsPath + "/" + datasetFileNames[3];

    /* Create MNIST dataset reader and setup paths */
    DatasetReader_MNIST<float> reader;
    reader.setTrainBatch(trainDatasetImages, trainDatasetLabels, trainDatasetObjectsNum);
    reader.setTestBatch(testDatasetImages, testDatasetLabels, testDatasetObjectsNum);
    reader.read();

    printf("Data loaded \n");

    printf("LeNet training started... \n");
    prediction::ModelPtr predictionModel = train(&reader);
    printf("LeNet training completed \n");

    printf("LeNet testing started \n");
    float top1ErrorRate = test(predictionModel, &reader);
    printf("LeNet testing completed \n");

    std::cout << "Top-1 error = " << top1ErrorRate * 100 << "%" << std::endl;
    return ((1.0f - top1ErrorRate) > top1ErrorRateThreshold) ? 0 : -1;
}

prediction::ModelPtr train(DatasetReader *reader)
{
    /* Fetch LeNet topology (configureNet defined in daal_lenet.h) */
    training::TopologyPtr topology = configureNet();

    /* Create the neural network training algorithm and set batch size and optimization solver */
    training::Batch<> net(getOptimizationSolver(learningRate));

    /* Fetch the training data and labels */
    TensorPtr trainingData        = reader->getTrainData();
    TensorPtr trainingGroundTruth = reader->getTrainGroundTruth();

    /* Initialize neural network with given topology */
    Collection<size_t> sampleSize = trainingData->getDimensions();
    sampleSize[0] = batchSize;
    net.initialize(sampleSize, *topology);

    /* Set the input data batch to the neural network */
    net.input.set(training::data, trainingData);

    /* Set the input ground truth (labels) batch to the neural network */
    net.input.set(training::groundTruth, trainingGroundTruth);

    for (size_t iter = 0; iter < maxIteraions; iter++)
    {
        /* Compute the neural network forward and backward passes and update */
        /* weights and biases according to the optimization solver */
        net.compute();
    }

    /* Get prediction model */
    training::ResultPtr trainingResult = net.getResult();
    training::ModelPtr trainedModel = trainingResult->get(training::model);
    return trainedModel->getPredictionModel<float>();
}

float test(prediction::ModelPtr predictionModel, DatasetReader *reader)
{
    /* Create the neural network prediction algorithm */
    prediction::Batch<> net;

    /* Fetch the testing data and labels */
    TensorPtr testingData         = reader->getTestData();
    TensorPtr testingGroundTruth  = reader->getTestGroundTruth();

    /* Set the prediction model retrieved from the training stage */
    net.input.set(prediction::model, predictionModel);

    /* Set the input ground truth (labels) batch to the neural network */
    net.input.set(prediction::data, testingData);

    /* Compute the neural network forward pass */
    net.compute();

    TensorPtr prediction = net.getResult()->get(prediction::prediction);

    /* Print first 100 predicted classes for each test sample */
    std::cout << "First 100 predictions ( predicted | ground truth ):" << std::endl;
    printPredictedClasses(prediction, testingGroundTruth, 100);

    /* Create auxiliary object to compute error rates (defined in services.h) */
    ClassificationErrorCounter errorRateCounter(prediction, testingGroundTruth);
    return errorRateCounter.getTop1ErrorRate();
}

SharedPtr<optimization_solver::adagrad::Batch<float> > getOptimizationSolver(float learningRate)
{
    /* Create 1 x 1 NumericTable to store learning rate */
    NumericTablePtr learningRateSequence = NumericTablePtr(new HomogenNumericTable<float>(
        1, 1, NumericTable::doAllocate, learningRate));

    /* Create AdaGrad optimization solver and set learning rate */
    optimization_solver::adagrad::Batch<float> *optalg = new optimization_solver::adagrad::Batch<float>();
    optalg->parameter.learningRate = learningRateSequence;
    optalg->parameter.batchSize = batchSize;
    optalg->parameter.nIterations = trainDatasetObjectsNum / batchSize;
    return SharedPtr<optimization_solver::adagrad::Batch<float> >(optalg);
}

File: samples/python/neural_networks/sources/daal_lenet.py
# file: daal_lenet.py
#===============================================================================
# Copyright 2017-2019 Intel Corporation.
#
# This software and the related documents are Intel copyrighted  materials,  and
# your use of  them is  governed by the  express license  under which  they were
# provided to you (License).  Unless the License provides otherwise, you may not
# use, modify, copy, publish, distribute,  disclose or transmit this software or
# the related documents without Intel's prior written permission.
#
# This software and the related documents  are provided as  is,  with no express
# or implied  warranties,  other  than those  that are  expressly stated  in the
# License.
#
# License:
# http://software.intel.com/en-us/articles/intel-sample-source-code-license-agr
# eement/
#===============================================================================

#
# !  Content:
# !    Python sample of LeNet training and prediction.
# !
# !*****************************************************************************

from __future__ import division, print_function

import os
import sys

import numpy as np

from daal.algorithms import optimization_solver
from daal.algorithms.neural_networks import training, prediction
from daal.algorithms.neural_networks import initializers
from daal.algorithms.neural_networks.layers import (
    convolution2d, pooling2d, maximum_pooling2d, fullyconnected, relu, loss
)
from daal.data_management import readOnly, SubtensorDescriptor, HomogenNumericTable

from image_dataset import DatasetReaderMNIST


def printPredictedClasses(_predictionResult, _testingGroundTruth):
    _prediction = _predictionResult.getResult(prediction.prediction)
    predictionDimensions = _prediction.getDimensions()

    predictionBlock = SubtensorDescriptor()
    _prediction.getSubtensor([], 0, predictionDimensions[0], readOnly, predictionBlock)
    predictionPtr = predictionBlock.getArray()

    testGroundTruthBlock = SubtensorDescriptor(ntype=np.intc)
    _testingGroundTruth.getSubtensor([], 0, predictionDimensions[0], readOnly, testGroundTruthBlock)
    testGroundTruthPtr = testGroundTruthBlock.getArray().flatten()

    # Print predicted classes
    maxPIndex = np.argmax(predictionPtr, axis=1)
    for i in range(predictionDimensions[0]):
        for p in predictionPtr[i]:
            print("{:.4f} ".format(p), end="")
        print(" -> {} | {}".format(maxPIndex[i], testGroundTruthPtr[i]))

    _prediction.releaseSubtensor(predictionBlock)
    _testingGroundTruth.releaseSubtensor(testGroundTruthBlock)


def configureNet():
    # Create convolution layer
    convolution1 = convolution2d.Batch()
    convolution1.parameter.kernelSizes = convolution2d.KernelSizes(3, 3)
    convolution1.parameter.strides = convolution2d.Strides(1, 1)
    convolution1.parameter.nKernels = 32
    convolution1.parameter.weightsInitializer = initializers.xavier.Batch()
    convolution1.parameter.biasesInitializer = initializers.uniform.Batch(0, 0)

    # Create pooling layer
    maxpooling1 = maximum_pooling2d.Batch(4)
    maxpooling1.parameter.kernelSizes = pooling2d.KernelSizes(2, 2)
    maxpooling1.parameter.paddings = pooling2d.Paddings(0, 0)
    maxpooling1.parameter.strides = pooling2d.Strides(2, 2)

    # Create convolution layer
    convolution2 = convolution2d.Batch()
    convolution2.parameter.kernelSizes = convolution2d.KernelSizes(5, 5)
    convolution2.parameter.strides = convolution2d.Strides(1, 1)
    convolution2.parameter.nKernels = 64
    convolution2.parameter.weightsInitializer = initializers.xavier.Batch()
    convolution2.parameter.biasesInitializer = initializers.uniform.Batch(0, 0)

    # Create pooling layer
    maxpooling2 = maximum_pooling2d.Batch(4)
    maxpooling2.parameter.kernelSizes = pooling2d.KernelSizes(2, 2)
    maxpooling2.parameter.paddings = pooling2d.Paddings(0, 0)
    maxpooling2.parameter.strides = pooling2d.Strides(2, 2)

    # Create fullyconnected layer
    fullyconnected3 = fullyconnected.Batch(256)
    fullyconnected3.parameter.weightsInitializer = initializers.xavier.Batch()
    fullyconnected3.parameter.biasesInitializer = initializers.uniform.Batch(0, 0)

    # Create ReLU layer
    relu3 = relu.Batch()

    # Create fully connected layer
    fullyconnected4 = fullyconnected.Batch(10)
    fullyconnected4.parameter.weightsInitializer = initializers.xavier.Batch()
    fullyconnected4.parameter.biasesInitializer = initializers.uniform.Batch(0, 0)

    # Create Softmax layer
    softmax = loss.softmax_cross.Batch()

    # Create LeNet Topology
    topology = training.Topology()
    conv1 = topology.add(convolution1)
    pool1 = topology.add(maxpooling1)
    topology.get(conv1).addNext(pool1)
    conv2 = topology.add(convolution2)
    topology.get(pool1).addNext(conv2)
    pool2 = topology.add(maxpooling2)
    topology.get(conv2).addNext(pool2)
    fc3 = topology.add(fullyconnected3)
    topology.get(pool2).addNext(fc3)
    r3 = topology.add(relu3)
    topology.get(fc3).addNext(r3)
    fc4 = topology.add(fullyconnected4)
    topology.get(r3).addNext(fc4)
    sm1 = topology.add(softmax)
    topology.get(fc4).addNext(sm1)

    return topology


# LeNet training
def train(trainingData, trainingGroundTruth):
    batchSize = 10
    learningRate = 0.01

    sgdAlgorithm = optimization_solver.sgd.Batch(fptype=np.float32)
    arr = np.array([[learningRate]], dtype=np.float32)
    sgdAlgorithm.parameter.learningRateSequence = HomogenNumericTable(arr, ntype=np.float32)
    sgdAlgorithm.parameter.batchSize = batchSize
    sgdAlgorithm.parameter.nIterations = int(trainingData.getDimensionSize(0) / sgdAlgorithm.parameter.batchSize)

    topology = configureNet()

    net = training.Batch(sgdAlgorithm)

    sampleSize = trainingData.getDimensions()
    sampleSize[0] = batchSize
    net.initialize(sampleSize, topology)

    net.input.setInput(training.data, trainingData)
    net.input.setInput(training.groundTruth, trainingGroundTruth)

    res = net.compute()

    return res.get(training.model).getPredictionModel_Float64()


# LeNet testing
def test(predictionModel, testingData, testingGroundTruth):
    net = prediction.Batch()

    net.input.setModelInput(prediction.model, predictionModel)
    net.input.setTensorInput(prediction.data, testingData)

    predictionResult = net.compute()

    printPredictedClasses(predictionResult, testingGroundTruth)

    return predictionResult


# check prediction results
def checkResult(predictionResult, testingGroundTruth, TestDataCount):
    pred = predictionResult.getResult(prediction.prediction)
    predictionDimensions = pred.getDimensions()

    predictionBlock = SubtensorDescriptor()
    pred.getSubtensor([], 0, predictionDimensions[0], readOnly, predictionBlock)
    predictionPtr = predictionBlock.getArray()

    testGroundTruthBlock = SubtensorDescriptor(ntype=np.intc)
    testingGroundTruth.getSubtensor([], 0, predictionDimensions[0], readOnly, testGroundTruthBlock)
    testGroundTruthPtr = testGroundTruthBlock.getArray().flatten()
    maxPIndex = 0
    trueCount = 0

    # validation accuracy finding
    maxPIndex = np.argmax(predictionPtr, axis=1)
    trueCount = np.sum(maxPIndex == testGroundTruthPtr)

    pred.releaseSubtensor(predictionBlock)
    testingGroundTruth.releaseSubtensor(testGroundTruthBlock)

    return True if trueCount / TestDataCount > 0.9 else False


def main():
    TrainDataCount = 50000
    TestDataCount = 100

    datasetFileNames = [
        os.path.join('data', 'train-images-idx3-ubyte'),
        os.path.join('data', 'train-labels-idx1-ubyte'),
        os.path.join('data', 't10k-images-idx3-ubyte'),
        os.path.join('data', 't10k-labels-idx1-ubyte')
    ]

    print("Data loading started... ")

    reader = DatasetReaderMNIST()
    reader.setTrainBatch(datasetFileNames[0], datasetFileNames[1], TrainDataCount)
    reader.setTestBatch(datasetFileNames[2], datasetFileNames[3], TestDataCount)
    reader.read()

    print("Data loaded ")

    trainingData = reader._trainData
    trainingGroundTruth = reader._trainGroundTruth
    testingData = reader._testData
    testingGroundTruth = reader._testGroundTruth

    print("LeNet training started... ")

    predictionModel = train(trainingData, trainingGroundTruth)

    print("LeNet training completed ")
    print("LeNet testing started ")

    predictionResult = test(predictionModel, testingData, testingGroundTruth)

    return 0 if checkResult(predictionResult, testingGroundTruth, TestDataCount) else -1

if __name__ == '__main__':
    sys.exit(main())
